source/link/linker.cpp

```diff
@@ -0,0 +1,292 @@
+// Copyright (c) 2017 Pierre Moreau
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "spirv-tools/linker.hpp"
+
+#include "spirv/1.2/spirv.hpp11"
+
+#include <cstring>
+#include <cstdio>
+
+#include <iostream>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+#include "opt/build_module.h"
+#include "opt/ir_loader.h"
+#include "opt/make_unique.h"
+#include "opt/module.h"
+#include "spirv_target_env.h"
+#include "spirv-tools/libspirv.hpp"
+
+
+namespace {
+
+
+} // anonymous namespace
+
+namespace spvtools {
+
+using namespace ir;
+
+// Structs for holding the data members for SpvTools.
+struct SpirvLinker::Impl {
+  explicit Impl(spv_target_env env) : context(spvContextCreate(env)) {
+    // The default consumer in spv_context_t is a null consumer, which provides
+    // equivalent functionality (from the user's perspective) as a real consumer
+    // does nothing.
+  }
+  ~Impl() { spvContextDestroy(context); }
+
+  spv_context context;  // C interface context object.
+};
+
+SpirvLinker::SpirvLinker(spv_target_env env) : impl_(new Impl(env)) {}
+
+SpirvLinker::~SpirvLinker() {}
+
+void SpirvLinker::SetMessageConsumer(MessageConsumer consumer) {
+  SetContextMessageConsumer(impl_->context, std::move(consumer));
+}
+
+bool SpirvLinker::Link(const std::vector<std::vector<uint32_t>>& binaries,
+                       std::vector<uint32_t>& linked_binary,
+                       uint32_t options) const {
+  auto linkedModule = MakeUnique<Module>();
+
+  std::vector<std::unique_ptr<Module>> modules;
+  modules.reserve(binaries.size());
+
+  uint32_t id_bound = 0u;
+  for (const auto& mod : binaries) {
+    std::unique_ptr<Module> module = BuildModule(impl_->context->target_env, impl_->context->consumer, mod.data(), mod.size());
+    if (module == nullptr)
+      return 1;
```

Comment:
Should be a failure case.
Should emit a diagnostic in failure cases.

We have a coding pattern of doing both diagnostic emission and returning of scalar result.  See use of diag() in source/val/validation_state.h, or uses of AssemblyContext::diagnostic in the assembler (source/text.cpp).